File: 은지/LV2/수식 최대화.js
function solution(expression) {
  var answer = 0;
  var stack_n = []
  var stack_s = []
  var num = ""
  for (let e of expression) {
    if (e === "-") {
      stack_n.push(parseInt(num))
      num = ""
      stack_s.push(e)
    } else if (e === "+") {
      stack_n.push(parseInt(num))
      num = ""
      stack_s.push(e)
    } else if (e === "*") {
      stack_n.push(parseInt(num))
      num = ""
      stack_s.push(e)
    } else {
      num += e
    }
  }
  stack_n.push(parseInt(num))

  var symbol = [["*", "+", "-"], ["*", "-", "+"], ["+", "*", "-"], ["+", "-", "*"], ["-", "+", "*"], ["-", "*", "+"]]
  for (let v of symbol) {
    var stack_ss = stack_s.slice(0, stack_s.length)
    var stack_nn = stack_n.slice(0, stack_n.length)

    for (let j=0; j<3; j++) {
      var i = 0
      while (i < stack_ss.length) {
        if (stack_nn.length === 1) {
          break
        }
        if (stack_ss[i] === v[j]) {
          var w = parseInt(stack_nn.splice(i+1, 1))
          stack_ss.splice(i, 1)
          if (v[j] === "*") {
            stack_nn[i] = parseInt(stack_nn[i] * w)
          } else if (v[j] === "+") {
            stack_nn[i] = parseInt(stack_nn[i] + w)
          } else if (v[j] === "-") {
            stack_nn[i] = parseInt(stack_nn[i] - w)
          }
        } else {
          i += 1
        }
      }
    }
    if (answer < Math.abs(stack_nn[0])) {
      answer = Math.abs(stack_nn[0])
    }
  }
  return answer;
}

e = "200-300-500-600*40+500+500"
console.log(solution(e))

File: 은지/LV2/수식 최대화.py
import itertools
import copy

def solution(expression):
    answer = 0
    # 숫자, 기호 정리
    stack_n = []
    stack_s = []
    num = ""
    for e in expression:
        if e == "-":
            stack_n.append(int(num))
            num = ""
            stack_s.append(e)
        elif e == "+":
            stack_n.append(int(num))
            num = ""
            stack_s.append(e)
        elif e == "*":
            stack_n.append(int(num))
            num = ""
            stack_s.append(e)
        else:
            num += e
    stack_n.append(int(num))

    # 모든 경우의 수 계산하며 최대값 탐색
    symbol = ["*", "+", "-"]
    for v in itertools.permutations(symbol, 3):
        stack_ss = copy.deepcopy(stack_s)
        stack_nn = copy.deepcopy(stack_n)
        for j in range(3):
            i = 0
            while i < len(stack_ss):
                if len(stack_nn) == 1:
                    break
                if stack_ss[i] == v[j]:
                    w = stack_nn.pop(i+1)
                    stack_ss.pop(i)
                    if v[j] == "*":
                        stack_nn[i] = stack_nn[i] * w
                    elif v[j] == "+":
                        stack_nn[i] = stack_nn[i] + w
                    elif v[j] == "-":
                        stack_nn[i] = stack_nn[i] - w
                else:
                    i += 1
        if answer < abs(stack_nn[0]):
            answer = abs(stack_nn[0])
    return answer

e = "100-200*300-500+20"
print(solution(e))